// allure/dist/commands/agent-human-report.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _AgentHumanReportPlugin_realtime, _AgentHumanReportPlugin_generated;
import * as console from "node:console";
import { readConfig, readRawConfig } from "@allurereport/core";
import AwesomePlugin from "@allurereport/plugin-awesome";
export const AGENT_HUMAN_REPORT_THRESHOLD = 1000;
const createInitialHumanReportStatus = (mode) => ({
    schema_version: "allure-agent-human-report/v1",
    mode,
    status: mode === "off" ? "disabled" : "pending",
    result_count: null,
    threshold: AGENT_HUMAN_REPORT_THRESHOLD,
    path: null,
    reports: [],
    reason: mode === "off" ? "disabled by --report off" : null,
    error: null,
});
const isAgentPluginId = (id) => id === "agent" || id === "plugin-agent";
const isAwesomeDescriptor = (id, descriptor) => id === "awesome" ||
    id === "@allurereport/plugin-awesome" ||
    descriptor.import === "awesome" ||
    descriptor.import === "@allurereport/plugin-awesome";
const configuredAwesomeOptions = async (cwd, configPath) => {
    const rawConfig = await readRawConfig(cwd, configPath);
    const plugins = rawConfig.plugins ?? {};
    for (const id of Object.keys(plugins)) {
        const descriptor = plugins[id];
        if (isAwesomeDescriptor(id, descriptor)) {
            return (descriptor.options ?? {});
        }
    }
    return {};
};
const recordGeneratedReport = (status, pluginId, path) => {
    const alreadyRecorded = status.reports.some((report) => report.plugin_id === pluginId && report.path === path);
    if (!alreadyRecorded) {
        status.reports.push({
            plugin_id: pluginId,
            path,
        });
    }
    status.status = status.error ? "failed" : "generated";
    status.path ?? (status.path = path);
    status.reason = null;
    status.generated_at = new Date().toISOString();
};
const recordFailedReport = (status, pluginId, error) => {
    const message = error instanceof Error ? error.message : String(error);
    status.status = "failed";
    status.error ?? (status.error = message);
    status.errors = [...(status.errors ?? []), { plugin_id: pluginId, message }];
};
class AgentHumanReportPlugin {
    constructor(params) {
        this.params = params;
        _AgentHumanReportPlugin_realtime.set(this, void 0);
        _AgentHumanReportPlugin_generated.set(this, false);
        this.start = async (_context, _store, realtime) => {
            __classPrivateFieldSet(this, _AgentHumanReportPlugin_realtime, realtime, "f");
        };
        this.done = async (context, store) => {
            const { mode, status, threshold, pluginId, plugin, reportPath } = this.params;
            const resultCount = (await store.allTestResults()).length;
            status.result_count = resultCount;
            if (status.status === "skipped") {
                return;
            }
            if (mode === "auto" && resultCount > threshold) {
                status.status = "skipped";
                status.reason = `result count ${resultCount} exceeds threshold ${threshold}`;
                return;
            }
            try {
                await plugin.start?.(context, store, __classPrivateFieldGet(this, _AgentHumanReportPlugin_realtime, "f"));
                await plugin.done?.(context, store);
                __classPrivateFieldSet(this, _AgentHumanReportPlugin_generated, true, "f");
                recordGeneratedReport(status, pluginId, reportPath);
            }
            catch (error) {
                recordFailedReport(status, pluginId, error);
                console.error(`agent human report ${pluginId} error`, error);
            }
        };
        this.info = async (context, store) => {
            if (!__classPrivateFieldGet(this, _AgentHumanReportPlugin_generated, "f")) {
                return undefined;
            }
            return this.params.plugin.info?.(context, store);
        };
    }
}
_AgentHumanReportPlugin_realtime = new WeakMap(), _AgentHumanReportPlugin_generated = new WeakMap();
const wrapHumanReportPlugin = (params) => {
    const { mode, status, plugin, reportPath = `${plugin.id}/` } = params;
    return {
        ...plugin,
        plugin: new AgentHumanReportPlugin({
            mode,
            status,
            threshold: AGENT_HUMAN_REPORT_THRESHOLD,
            pluginId: plugin.id,
            plugin: plugin.plugin,
            reportPath,
        }),
    };
};
export const createAgentHumanReportConfig = async ({ mode, cwd, configPath, outputDir, configOverride, }) => {
    const status = createInitialHumanReportStatus(mode);
    const statusProvider = () => status;
    if (mode === "off") {
        return {
            status,
            statusProvider,
            plugins: [],
        };
    }
    if (mode === "auto" || mode === "awesome") {
        const awesomeOptions = {
            ...(await configuredAwesomeOptions(cwd, configPath)),
            singleFile: true,
        };
        const awesomePlugin = {
            id: "awesome",
            enabled: true,
            options: awesomeOptions,
            plugin: new AwesomePlugin(awesomeOptions),
        };
        return {
            status,
            statusProvider,
            plugins: [
                wrapHumanReportPlugin({
                    mode,
                    status,
                    plugin: awesomePlugin,
                    reportPath: "awesome/index.html",
                }),
            ],
        };
    }
    const reportConfig = await readConfig(cwd, configPath, {
        ...configOverride,
        output: outputDir,
    });
    const plugins = reportConfig.plugins
        ?.filter((plugin) => !isAgentPluginId(plugin.id))
        .map((plugin) => wrapHumanReportPlugin({
        mode,
        status,
        plugin,
    })) ?? [];
    if (!plugins.length) {
        status.status = "disabled";
        status.reason = "no configured non-agent report plugins";
    }
    return {
        status,
        statusProvider,
        plugins,
    };
};
