// allure/dist/commands/classic.js
import * as console from "node:console";
import { realpath } from "node:fs/promises";
import process, { exit } from "node:process";
import { AllureReport, readConfig } from "@allurereport/core";
import ClassicPlugin from "@allurereport/plugin-classic";
import { Command, Option } from "clipanion";
import { red } from "yoctocolors";
import { findAllureResultDirectories } from "../utils/fileSystem.js";
export class ClassicCommand extends Command {
    constructor() {
        super(...arguments);
        this.resultsDir = Option.Rest({
            name: "Patterns to match test results directories in the current working directory (default: ./**/allure-results)",
        });
        this.config = Option.String("--config,-c", {
            description: "The path Allure config file",
        });
        this.cwd = Option.String("--cwd", {
            description: "The working directory for the command to run (default: current working directory)",
        });
        this.output = Option.String("--output,-o", {
            description: "The output directory name. Absolute paths are accepted as well",
        });
        this.reportName = Option.String("--report-name,--name", {
            description: "The report name",
        });
        this.reportLanguage = Option.String("--report-language,--lang", {
            description: "Default language of the report (default: OS language)",
        });
        this.singleFile = Option.Boolean("--single-file", {
            description: "Generate single file report",
        });
        this.historyPath = Option.String("--history-path,-h", {
            description: "The path to history file",
        });
        this.knownIssues = Option.String("--known-issues", {
            description: "Path to the known issues file. Updates the file and quarantines failed tests when specified",
        });
    }
    async execute() {
        const cwd = await realpath(this.cwd ?? process.cwd());
        const { resultDirectories, patterns } = await findAllureResultDirectories(cwd, this.resultsDir);
        if (!resultDirectories.length) {
            console.error(red(`No test results directories found matching pattern: ${patterns}`));
            exit(1);
            return;
        }
        const before = new Date().getTime();
        const defaultClassicOptions = {
            singleFile: this.singleFile ?? false,
            reportLanguage: this.reportLanguage,
        };
        const config = await readConfig(cwd, this.config, {
            output: this.output,
            name: this.reportName,
            knownIssuesPath: this.knownIssues,
            historyPath: this.historyPath,
        });
        config.plugins = [
            {
                id: "classic",
                enabled: true,
                options: defaultClassicOptions,
                plugin: new ClassicPlugin(defaultClassicOptions),
            },
        ];
        const allureReport = new AllureReport(config);
        await allureReport.start();
        for (const directory of resultDirectories) {
            await allureReport.readDirectory(directory);
        }
        await allureReport.done();
        const after = new Date().getTime();
        console.log(`the report successfully generated (${after - before}ms)`);
    }
}
ClassicCommand.paths = [["classic"]];
ClassicCommand.usage = Command.Usage({
    category: "Reports",
    description: "Generates Allure Classic report based on provided Allure Results",
    details: "This command generates an Allure Classic report from the provided Allure Results directory.",
    examples: [
        ["classic ./allure-results", "Generate a report from the ./allure-results directory"],
        [
            "classic ./allure-results --output custom-report",
            "Generate a report from the ./allure-results directory to the custom-report directory",
        ],
        [
            "classic ./packages/*/allure-results",
            "Generate a report from all Allure result directories matching the pattern",
        ],
        [
            "classic ./packages/foo/allure-results ./packages/bar/allure-results",
            "Generate a report from two Allure result directories",
        ],
    ],
});
