// @allurereport/plugin-api/dist/utils/tree.js
import { emptyStatistic } from "@allurereport/core-api";
import { md5 } from "./misc.js";
const addLeaf = (childLeaves, node, nodeId) => {
    if (node.leaves === undefined) {
        node.leaves = [];
    }
    let leaves = childLeaves.get(node);
    if (!leaves) {
        leaves = new Set(node.leaves);
        childLeaves.set(node, leaves);
    }
    if (leaves.has(nodeId)) {
        return;
    }
    leaves.add(nodeId);
    node.leaves.push(nodeId);
};
const addGroup = (childGroups, node, nodeId) => {
    if (node.groups === undefined) {
        node.groups = [];
    }
    let groups = childGroups.get(node);
    if (!groups) {
        groups = new Set(node.groups);
        childGroups.set(node, groups);
    }
    if (groups.has(nodeId)) {
        return;
    }
    groups.add(nodeId);
    node.groups.push(nodeId);
};
const createTree = (data, classifier, leafFactory, groupFactory, addLeafToGroup = () => { }) => {
    const groupsByClassifier = new Map();
    const leavesById = {};
    const groupsById = {};
    const childLeaves = new WeakMap();
    const childGroups = new WeakMap();
    const root = { groups: [], leaves: [] };
    for (const item of data) {
        const leaf = leafFactory(item);
        leavesById[leaf.nodeId] = leaf;
        const itemGroups = classifier(item);
        let parentGroups = [root];
        for (const layer of itemGroups) {
            if (layer.length === 0) {
                break;
            }
            const nextParentGroups = [];
            const nextParentGroupIds = new Set();
            for (const group of layer) {
                for (const parentGroup of parentGroups) {
                    const parentId = "nodeId" in parentGroup ? parentGroup.nodeId : "";
                    let groupsByParent = groupsByClassifier.get(parentId);
                    if (groupsByParent === undefined) {
                        groupsByParent = new Map();
                        groupsByClassifier.set(parentId, groupsByParent);
                    }
                    if (groupsByParent.get(group) === undefined) {
                        const newGroup = groupFactory(parentId, group);
                        if (!newGroup || typeof newGroup !== "object" || typeof newGroup.nodeId !== "string") {
                            continue;
                        }
                        groupsByParent.set(group, newGroup);
                        groupsById[newGroup.nodeId] = newGroup;
                    }
                    const currentGroup = groupsByParent.get(group);
                    if (!currentGroup || typeof currentGroup !== "object") {
                        continue;
                    }
                    addGroup(childGroups, parentGroup, currentGroup.nodeId);
                    addLeafToGroup(currentGroup, leaf);
                    if (!nextParentGroupIds.has(currentGroup.nodeId)) {
                        nextParentGroupIds.add(currentGroup.nodeId);
                        nextParentGroups.push(currentGroup);
                    }
                }
            }
            parentGroups = nextParentGroups;
        }
        parentGroups.forEach((parentGroup) => {
            addLeaf(childLeaves, parentGroup, leaf.nodeId);
        });
    }
    return {
        root,
        groupsById,
        leavesById,
    };
};
const byLabelsWithFallback = (item, labelNames, fallbackValue) => {
    const labelsByName = new Map();
    for (const label of item.labels) {
        const values = labelsByName.get(label.name) ?? [];
        values.push(label.value ?? fallbackValue);
        labelsByName.set(label.name, values);
    }
    return labelNames.map((labelName) => labelsByName.get(labelName) ?? []).filter((layer) => layer.length > 0);
};
export const byLabels = (item, labelNames) => byLabelsWithFallback(item, labelNames, "__unknown");
export const filterTreeLabels = (data, labelNames) => {
    const requested = new Set(labelNames);
    const found = new Set();
    for (const item of data) {
        for (const label of item.labels) {
            if (requested.has(label.name)) {
                found.add(label.name);
            }
        }
        if (found.size === requested.size) {
            break;
        }
    }
    return labelNames.filter((labelName) => found.has(labelName));
};
export const createTreeByLabels = (data, labelNames, leafFactory, groupFactory, addLeafToGroup = () => { }) => {
    const leafFactoryFn = leafFactory ??
        ((tr) => {
            const { id, name, status, duration } = tr;
            return {
                nodeId: id,
                name,
                status,
                duration,
            };
        });
    const groupFactoryFn = groupFactory ??
        ((parentId, groupClassifier) => ({
            nodeId: md5((parentId ? `${parentId}.` : "") + groupClassifier),
            name: groupClassifier,
            statistic: emptyStatistic(),
        }));
    return createTree(data, (item) => byLabels(item, labelNames), leafFactoryFn, groupFactoryFn, addLeafToGroup);
};
export const createTreeByCategories = (data, leafFactory, groupFactory, addLeafToGroup = () => { }) => {
    const leafFactoryFn = leafFactory ??
        ((tr) => {
            const { id, name, status, duration } = tr;
            return {
                nodeId: id,
                name,
                status,
                duration,
            };
        });
    const groupFactoryFn = groupFactory ??
        ((parentId, groupClassifier) => ({
            nodeId: md5((parentId ? `${parentId}.` : "") + groupClassifier),
            name: groupClassifier,
            statistic: emptyStatistic(),
        }));
    return createTree(data, (item) => byCategories(item), leafFactoryFn, groupFactoryFn, addLeafToGroup);
};
export const byCategories = (item) => {
    const categories = item.categories ?? [];
    const result = [];
    for (const category of categories) {
        result.push([category.name]);
    }
    if (item.error?.message) {
        result.push([item.error?.message]);
    }
    return result;
};
export const preciseTreeLabels = (labelNames, trs, labelNamesAccessor = (tr) => tr.labels.map(({ name }) => name)) => {
    const requested = new Set(labelNames);
    const found = new Set();
    for (const tr of trs) {
        for (const labelName of labelNamesAccessor(tr)) {
            if (requested.has(labelName)) {
                found.add(labelName);
            }
        }
        if (found.size === requested.size) {
            break;
        }
    }
    const emitted = new Set();
    return labelNames.filter((labelName) => {
        if (!found.has(labelName) || emitted.has(labelName)) {
            return false;
        }
        emitted.add(labelName);
        return true;
    });
};
export const processTree = (tree, options) => {
    const visitedGroups = new Set();
    const { root, leavesById, groupsById } = tree;
    const processGroup = (group) => {
        if (group.groups?.length) {
            group.groups.forEach((groupId) => {
                const subGroup = groupsById[groupId];
                if (!subGroup || visitedGroups.has(groupId)) {
                    return;
                }
                processGroup(subGroup);
                visitedGroups.add(groupId);
            });
        }
        if (group.leaves?.length) {
            if (options.filter) {
                group.leaves = group.leaves.filter((leaveId) => options.filter(leavesById[leaveId]));
            }
            if (options.sort) {
                group.leaves = group.leaves.sort((a, b) => {
                    const leafA = leavesById[a];
                    const leafB = leavesById[b];
                    return options.sort(leafA, leafB);
                });
            }
            if (options.transform) {
                group.leaves.forEach((leafId, i) => {
                    leavesById[leafId] = options.transform(leavesById[leafId], i);
                });
            }
        }
        return group;
    };
    processGroup(root);
    return tree;
};
export const filterTree = (tree, predicate) => {
    return processTree(tree, { filter: predicate });
};
export const sortTree = (tree, comparator) => {
    return processTree(tree, { sort: comparator });
};
export const transformTree = (tree, transformer) => {
    return processTree(tree, { transform: transformer });
};
export const createTreeByTitlePath = (data, leafFactory, groupFactory, addLeafToGroup = () => { }) => {
    const leafFactoryFn = leafFactory ??
        ((tr) => {
            const { id, name, status, duration } = tr;
            return {
                nodeId: id,
                name,
                status,
                duration,
            };
        });
    const groupFactoryFn = groupFactory ??
        ((parentId, groupClassifier) => ({
            nodeId: md5((parentId ? `${parentId}.` : "") + groupClassifier),
            name: groupClassifier,
            statistic: emptyStatistic(),
        }));
    return createTree(data, (item) => (item.titlePath ?? []).map((segment) => [segment]), leafFactoryFn, groupFactoryFn, addLeafToGroup);
};
const byLabelsAndTitlePath = (item, labelNames) => {
    const leaves = byLabelsWithFallback(item, labelNames, "");
    const titlePath = item.titlePath;
    if (Array.isArray(titlePath) && titlePath.length > 0) {
        for (const segment of titlePath) {
            leaves.push([segment]);
        }
    }
    return leaves;
};
export const createTreeByLabelsAndTitlePath = (data, labelNames = [], leafFactory, groupFactory, addLeafToGroup = () => { }) => {
    const leafFactoryFn = leafFactory ??
        ((tr) => {
            const { id, name, status, duration } = tr;
            return {
                nodeId: id,
                name,
                status,
                duration,
            };
        });
    const groupFactoryFn = groupFactory ??
        ((parentId, groupClassifier) => ({
            nodeId: md5((parentId ? `${parentId}.` : "") + groupClassifier),
            name: groupClassifier,
            statistic: emptyStatistic(),
        }));
    return createTree(data, (item) => byLabelsAndTitlePath(item, labelNames), leafFactoryFn, groupFactoryFn, addLeafToGroup);
};
