// @allurereport/ci/dist/helpers/gitProvider.js
import { GitProvider } from "@allurereport/core-api";
const stripGitSuffix = (url) => url.replace(/\.git\/?$/i, "");
export const stripRefsHeads = (ref) => ref.replace(/^refs\/heads\//, "");
export const parsePullRequestNumberFromUrl = (pullRequestUrl) => {
    const trimmed = pullRequestUrl.trim();
    if (!trimmed) {
        return undefined;
    }
    const match = trimmed.match(/\/pull\/(\d+)\/?$/i) ??
        trimmed.match(/\/merge_requests\/(\d+)\/?$/i) ??
        trimmed.match(/\/pull-requests\/(\d+)\/?$/i);
    return match?.[1];
};
export const hostMatchesProvider = (host, provider) => {
    const normalizedHost = host.toLowerCase();
    switch (provider) {
        case GitProvider.Github:
            return (normalizedHost === "github.com" ||
                normalizedHost.endsWith(".github.com") ||
                normalizedHost.startsWith("github."));
        case GitProvider.Gitlab:
            return (normalizedHost === "gitlab.com" ||
                normalizedHost.endsWith(".gitlab.com") ||
                normalizedHost.startsWith("gitlab."));
        case GitProvider.Bitbucket:
            return (normalizedHost === "bitbucket.org" ||
                normalizedHost.endsWith(".bitbucket.org") ||
                normalizedHost.startsWith("bitbucket."));
        default:
            return false;
    }
};
export const normalizeGitRemoteUrl = (remote) => {
    const trimmed = remote.trim();
    if (!trimmed) {
        return undefined;
    }
    if (trimmed.startsWith("http://") || trimmed.startsWith("https://")) {
        return trimmed;
    }
    if (trimmed.startsWith("ssh://")) {
        try {
            const parsed = new URL(trimmed);
            return `https://${parsed.hostname}${parsed.pathname}`;
        }
        catch {
            return undefined;
        }
    }
    const scpMatch = /^[^@]+@([^:]+):(.+)$/.exec(trimmed);
    if (scpMatch) {
        return `https://${scpMatch[1]}/${scpMatch[2].replace(/^\//, "")}`;
    }
    return undefined;
};
export const inferGitProviderFromUrl = (remote) => {
    const normalized = normalizeGitRemoteUrl(remote);
    if (!normalized) {
        return undefined;
    }
    try {
        const host = new URL(normalized).hostname.toLowerCase();
        for (const provider of [GitProvider.Github, GitProvider.Gitlab, GitProvider.Bitbucket]) {
            if (hostMatchesProvider(host, provider)) {
                return provider;
            }
        }
    }
    catch {
        return undefined;
    }
    return undefined;
};
export const parseRepositorySlugFromUrl = (remote, provider) => {
    const normalized = normalizeGitRemoteUrl(remote);
    if (!normalized) {
        return undefined;
    }
    try {
        const pathname = new URL(normalized).pathname.replace(/^\/+/, "").replace(/\.git\/?$/i, "");
        if (!pathname) {
            return undefined;
        }
        const segments = pathname.split("/").filter(Boolean);
        if (provider === GitProvider.Github || provider === GitProvider.Bitbucket) {
            if (segments.length < 2) {
                return undefined;
            }
            return `${segments[0]}/${segments[1]}`;
        }
        return pathname;
    }
    catch {
        return undefined;
    }
};
export const resolveRepositoryFromGitUrl = (remote) => {
    const provider = inferGitProviderFromUrl(remote);
    if (!provider) {
        return undefined;
    }
    const slug = parseRepositorySlugFromUrl(remote, provider);
    if (!slug) {
        return undefined;
    }
    const normalized = normalizeGitRemoteUrl(remote);
    if (!normalized) {
        return undefined;
    }
    return {
        provider,
        slug,
        url: stripGitSuffix(normalized),
    };
};
