// @allurereport/core/dist/utils/perf.js
import { mkdir, writeFile } from "node:fs/promises";
import { join } from "node:path";
import { performance } from "node:perf_hooks";
export const PERF_METRICS_FILE = "allure-perf-metrics.json";
const MARK_PREFIX = "allure:perf:";
const ENABLED_VALUES = new Set(["1", "true", "yes", "on"]);
const SPANS = [];
const MARKS = new Set();
const MEASURES = new Set();
let sequence = 0;
const round = (value) => Number(value.toFixed(3));
export const PERF_METRIC_NAMES = {
    restoreStateTotal: "restoreState.total",
    restoreStateDump: "restoreState.dump",
    restoreStateAttachments: "restoreState.attachments",
    restoreStateStoreRestore: "restoreState.storeRestore",
    generateTotal: "generate.total",
    generateReadResults: "generate.readResults",
    generatePluginsDone: "generate.plugins.done",
    publishUploadTotal: "publish.upload.total",
    summaryGenerate: "summary.generate",
};
export const PERF_METRIC_PREFIXES = {
    generatePluginDone: "generate.plugin.done.",
    publishUploadPlugin: "publish.upload.plugin.",
};
export const isPerfMetricsEnabled = () => ENABLED_VALUES.has((process.env.ALLURE_PERF_METRICS ?? "").toLowerCase());
export const startPerfSpan = (name) => {
    if (!isPerfMetricsEnabled()) {
        return () => { };
    }
    const id = `${MARK_PREFIX}${sequence++}:${name}`;
    const startMark = `${id}:start`;
    const endMark = `${id}:end`;
    let ended = false;
    MARKS.add(startMark);
    MARKS.add(endMark);
    MEASURES.add(id);
    performance.mark(startMark);
    return () => {
        if (ended) {
            return;
        }
        ended = true;
        performance.mark(endMark);
        performance.measure(id, startMark, endMark);
        const entry = performance.getEntriesByName(id, "measure").at(-1);
        if (entry) {
            SPANS.push({
                name,
                startTimeMs: round(entry.startTime),
                durationMs: round(entry.duration),
            });
        }
        performance.clearMarks(startMark);
        performance.clearMarks(endMark);
        performance.clearMeasures(id);
        MARKS.delete(startMark);
        MARKS.delete(endMark);
        MEASURES.delete(id);
    };
};
export const measurePerf = async (name, fn) => {
    if (!isPerfMetricsEnabled()) {
        return fn();
    }
    const end = startPerfSpan(name);
    try {
        return await fn();
    }
    finally {
        end();
    }
};
export const getPerfMetricsPayload = () => {
    const byName = new Map();
    for (const span of SPANS) {
        const current = byName.get(span.name) ?? [];
        current.push(span);
        byName.set(span.name, current);
    }
    return {
        version: 1,
        generatedAt: new Date().toISOString(),
        timeOriginMs: round(performance.timeOrigin),
        spans: [...SPANS],
        summary: [...byName.entries()].map(([name, group]) => {
            const durations = group.map(({ durationMs }) => durationMs);
            const totalMs = durations.reduce((acc, duration) => acc + duration, 0);
            return {
                name,
                count: group.length,
                totalMs: round(totalMs),
                minMs: round(Math.min(...durations)),
                maxMs: round(Math.max(...durations)),
                avgMs: round(totalMs / group.length),
            };
        }),
    };
};
export const writePerfMetrics = async (output) => {
    if (!isPerfMetricsEnabled()) {
        return false;
    }
    const payload = getPerfMetricsPayload();
    await mkdir(output, { recursive: true });
    await writeFile(join(output, PERF_METRICS_FILE), `${JSON.stringify(payload, null, 2)}\n`, "utf8");
    resetPerfMetrics();
    return true;
};
export const resetPerfMetrics = () => {
    SPANS.length = 0;
    sequence = 0;
    for (const mark of MARKS) {
        performance.clearMarks(mark);
    }
    for (const measure of MEASURES) {
        performance.clearMeasures(measure);
    }
    MARKS.clear();
    MEASURES.clear();
};
