// @allurereport/web-commons/dist/subtreeExpansion.js
const isFailedOrBrokenNode = (statistic) => statistic === undefined || Boolean(statistic?.failed || statistic?.broken);
const getDefaultOpenedState = (statistic, root = false) => root || isFailedOrBrokenNode(statistic);
export const hasExpandableTreeChildren = (tree) => tree.trees.length > 0 || tree.leaves.length > 0;
export const collectExpandableSubtreeNodes = (tree) => {
    const nodes = [];
    const stack = [{ tree, isRoot: true }];
    while (stack.length > 0) {
        const current = stack.pop();
        if (!current) {
            continue;
        }
        nodes.push({
            id: current.tree.nodeId,
            openedByDefault: getDefaultOpenedState(current.tree.statistic),
            isRoot: current.isRoot,
        });
        current.tree.trees.forEach((nestedSubtree) => stack.push({ tree: nestedSubtree, isRoot: false }));
    }
    return nodes;
};
export const applySubtreeToggleState = (expandableSubtreeNodes, state, options) => {
    expandableSubtreeNodes.forEach((node) => {
        const shouldOpenSubtree = state === "all" ? true : state === "first" ? node.isRoot : false;
        const scopedId = options.toScopedId(node.id);
        const currentlyOpened = options.isOpened(scopedId, node.openedByDefault);
        if (currentlyOpened !== shouldOpenSubtree) {
            options.setOpened(scopedId, shouldOpenSubtree, node.openedByDefault);
        }
    });
};
