// @allurereport/web-commons/dist/treeSubtreeToggle.js
export const isSubtreeFirstLevelOnlyOpened = (rootId, rootOpenedByDefault, subtreeNodes, isNodeOpened) => {
    if (!isNodeOpened(rootId, rootOpenedByDefault)) {
        return false;
    }
    return subtreeNodes.filter((node) => !node.isRoot).every((node) => !isNodeOpened(node.id, node.openedByDefault));
};
export const isSubtreeExpandedAll = (subtreeNodes, isNodeOpened) => subtreeNodes.every((node) => isNodeOpened(node.id, node.openedByDefault));
export const getNextSubtreeToggleState = (payload) => {
    const { hasOnlyLeafResults, isSubtreeCollapsedAll, isSubtreeFirstLevelOnly, isSubtreeExpandedAll, lastSubtreeToggle, } = payload;
    if (hasOnlyLeafResults) {
        return isSubtreeCollapsedAll ? "all" : "none";
    }
    if (isSubtreeCollapsedAll) {
        return "first";
    }
    if (isSubtreeFirstLevelOnly) {
        return lastSubtreeToggle === "all" ? "none" : "all";
    }
    if (isSubtreeExpandedAll) {
        return "first";
    }
    return "all";
};
export const resolveNextSubtreeToggleState = (subtreeNodes, isNodeOpened, lastSubtreeToggle) => {
    const root = subtreeNodes.find((node) => node.isRoot);
    if (!root || subtreeNodes.length === 0) {
        return { nextState: "first", nextLastToggle: null };
    }
    const isSubtreeCollapsedAll = !isNodeOpened(root.id, root.openedByDefault);
    const isSubtreeFirstLevelOnly = isSubtreeFirstLevelOnlyOpened(root.id, root.openedByDefault, subtreeNodes, isNodeOpened);
    const subtreeExpandedAll = isSubtreeExpandedAll(subtreeNodes, isNodeOpened);
    const hasOnlyLeafResults = subtreeNodes.every((node) => node.isRoot);
    const nextState = getNextSubtreeToggleState({
        hasOnlyLeafResults,
        isSubtreeCollapsedAll,
        isSubtreeFirstLevelOnly,
        isSubtreeExpandedAll: subtreeExpandedAll,
        lastSubtreeToggle,
    });
    return {
        nextState,
        nextLastToggle: nextState !== "first" ? nextState : lastSubtreeToggle,
    };
};
export const getSubtreeToggleIcon = (payload) => {
    const { hasOnlyLeafResults, isSubtreeCollapsedAll, isSubtreeFirstLevelOnly } = payload;
    if (hasOnlyLeafResults) {
        return isSubtreeCollapsedAll ? "single-down" : "single-up";
    }
    if (isSubtreeCollapsedAll) {
        return "single-down";
    }
    if (isSubtreeFirstLevelOnly) {
        return "double-down";
    }
    return "double-up";
};
