// @allurereport/service/dist/testops.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _AllureTestOpsClient_instances, _AllureTestOpsClient_url, _AllureTestOpsClient_accessToken, _AllureTestOpsClient_projectId, _AllureTestOpsClient_client, _AllureTestOpsClient_authorizedClient;
import { readFile } from "node:fs/promises";
import { extname, join as joinPosix } from "node:path/posix";
import { ALLURE_SERVICE_TESTOPS_PREFIX, } from "./model.js";
import { createServiceHttpClient, uploadReport } from "./utils/http.js";
import { parseServiceToken } from "./utils/token.js";
const DEFAULT_UPLOAD_CONTENT_TYPE = "application/octet-stream";
const CONTENT_TYPES = {
    ".css": "text/css",
    ".gif": "image/gif",
    ".html": "text/html",
    ".jpeg": "image/jpeg",
    ".jpg": "image/jpeg",
    ".js": "application/javascript",
    ".json": "application/json",
    ".map": "application/json",
    ".mjs": "application/javascript",
    ".otf": "font/otf",
    ".png": "image/png",
    ".svg": "image/svg+xml",
    ".ttf": "font/ttf",
    ".webp": "image/webp",
    ".woff": "font/woff",
    ".woff2": "font/woff2",
};
const contentTypeByFilename = (filename) => CONTENT_TYPES[extname(filename)] ?? DEFAULT_UPLOAD_CONTENT_TYPE;
const createUploadBlob = (content, filename) => new Blob([content], { type: contentTypeByFilename(filename) });
const createReportFileUrl = (baseUrl, reportUuid, reportFilename) => `${baseUrl}/api/test-report/view/${joinPosix(reportUuid, reportFilename)}`;
export class AllureTestOpsClient {
    constructor(config) {
        _AllureTestOpsClient_instances.add(this);
        this.config = config;
        _AllureTestOpsClient_url.set(this, void 0);
        _AllureTestOpsClient_accessToken.set(this, void 0);
        _AllureTestOpsClient_projectId.set(this, void 0);
        _AllureTestOpsClient_client.set(this, void 0);
        if (!config?.accessToken) {
            throw new Error("Allure TestOps access token is required");
        }
        if (!config.accessToken.startsWith(ALLURE_SERVICE_TESTOPS_PREFIX)) {
            throw new Error("Allure service access token is invalid");
        }
        const accessTokenPayload = parseServiceToken(config.accessToken);
        const projectId = Number(accessTokenPayload.projectId);
        if (!Number.isFinite(projectId)) {
            throw new Error("Given access token doesn't contain project id");
        }
        if (!accessTokenPayload.url) {
            throw new Error("Given access token doesn't contain url");
        }
        __classPrivateFieldSet(this, _AllureTestOpsClient_accessToken, accessTokenPayload.accessToken, "f");
        __classPrivateFieldSet(this, _AllureTestOpsClient_projectId, projectId, "f");
        __classPrivateFieldSet(this, _AllureTestOpsClient_url, accessTokenPayload.url.replace(/\/$/, ""), "f");
    }
    async downloadHistory() {
        return [];
    }
    async createReport(payload) {
        const client = await __classPrivateFieldGet(this, _AllureTestOpsClient_instances, "m", _AllureTestOpsClient_authorizedClient).call(this);
        const { reportName, reportUuid } = payload;
        const { url } = await client.post("/api/test-report", {
            body: {
                projectId: __classPrivateFieldGet(this, _AllureTestOpsClient_projectId, "f"),
                isPublic: !this.config?.private,
                reportName,
                reportUuid,
            },
        });
        return new URL(url, __classPrivateFieldGet(this, _AllureTestOpsClient_url, "f"));
    }
    async completeReport(payload) {
        const client = await __classPrivateFieldGet(this, _AllureTestOpsClient_instances, "m", _AllureTestOpsClient_authorizedClient).call(this);
        const { reportUuid } = payload;
        return client.post(`/api/test-report/${reportUuid}/complete`);
    }
    async deleteReport(payload) {
        const client = await __classPrivateFieldGet(this, _AllureTestOpsClient_instances, "m", _AllureTestOpsClient_authorizedClient).call(this);
        const { reportUuid } = payload;
        return client.delete(`/api/test-report/${reportUuid}`);
    }
    async addReportAsset(payload) {
        const { filename, file, filepath, signal } = payload;
        if (!file && !filepath) {
            throw new Error("File or filepath is required");
        }
        let content = file;
        if (!content) {
            content = signal ? await readFile(filepath, { signal }) : await readFile(filepath);
        }
        const form = new FormData();
        form.set("filename", filename);
        form.set("file", createUploadBlob(content, filename), filename);
        const client = await __classPrivateFieldGet(this, _AllureTestOpsClient_instances, "m", _AllureTestOpsClient_authorizedClient).call(this);
        return client.post("/api/test-report/upload", {
            body: form,
            headers: {
                "Content-Type": "multipart/form-data",
            },
            ...(signal ? { signal } : {}),
        });
    }
    async addReportFile(payload) {
        const { reportUuid, filename, file, filepath, pluginId, signal } = payload;
        const reportFilename = pluginId ? joinPosix(pluginId, filename) : filename;
        if (!file && !filepath) {
            throw new Error("File or filepath is required");
        }
        let content = file;
        if (!content) {
            content = signal ? await readFile(filepath, { signal }) : await readFile(filepath);
        }
        const form = new FormData();
        form.set("filename", reportFilename);
        form.set("file", createUploadBlob(content, reportFilename), reportFilename);
        const client = await __classPrivateFieldGet(this, _AllureTestOpsClient_instances, "m", _AllureTestOpsClient_authorizedClient).call(this);
        await client.post(`/api/test-report/${reportUuid}/upload`, {
            body: form,
            headers: {
                "Content-Type": "multipart/form-data",
            },
            ...(signal ? { signal } : {}),
        });
        return createReportFileUrl(__classPrivateFieldGet(this, _AllureTestOpsClient_url, "f"), reportUuid, reportFilename);
    }
    async uploadReport(payload) {
        return uploadReport({
            ...payload,
            uploadConcurrency: this.config.uploadConcurrency,
            uploadMaxAttempts: this.config.uploadMaxAttempts,
            uploadMaxSimultaneousFailures: this.config.uploadMaxSimultaneousFailures,
            addReportAsset: this.addReportAsset.bind(this),
            addReportFile: this.addReportFile.bind(this),
        });
    }
}
_AllureTestOpsClient_url = new WeakMap(), _AllureTestOpsClient_accessToken = new WeakMap(), _AllureTestOpsClient_projectId = new WeakMap(), _AllureTestOpsClient_client = new WeakMap(), _AllureTestOpsClient_instances = new WeakSet(), _AllureTestOpsClient_authorizedClient = async function _AllureTestOpsClient_authorizedClient() {
    if (__classPrivateFieldGet(this, _AllureTestOpsClient_client, "f")) {
        return __classPrivateFieldGet(this, _AllureTestOpsClient_client, "f");
    }
    __classPrivateFieldSet(this, _AllureTestOpsClient_client, createServiceHttpClient(__classPrivateFieldGet(this, _AllureTestOpsClient_url, "f"), {
        apiToken: __classPrivateFieldGet(this, _AllureTestOpsClient_accessToken, "f"),
    }), "f");
    return __classPrivateFieldGet(this, _AllureTestOpsClient_client, "f");
};
