// @allurereport/core/dist/store/retrySubstore.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _RetrySubstore_instances, _RetrySubstore_testResultsByRetryHash, _RetrySubstore_testResultIngestIndexById, _RetrySubstore_getIngestIndex, _RetrySubstore_compareResults;
import { md5 } from "../utils/crypto.js";
const NO_RETRIES = [];
const compareByNameThenValue = (first, second) => first.name.localeCompare(second.name) || first.value.localeCompare(second.value);
const stringifyRetryParameters = (parameters = []) => parameters
    .filter((parameter) => !parameter.excluded)
    .sort(compareByNameThenValue)
    .map((parameter) => `${parameter.name}:${parameter.value}`)
    .join(",");
export const calculateParametersHash = (parameters = []) => md5(stringifyRetryParameters(parameters));
export const calculateRetryHash = (testCaseId, parametersHash, environmentId) => {
    if (!testCaseId) {
        return undefined;
    }
    return md5(`${testCaseId}:${parametersHash}:${environmentId ?? "default"}`);
};
export class RetrySubstore {
    constructor() {
        _RetrySubstore_instances.add(this);
        _RetrySubstore_testResultsByRetryHash.set(this, new Map());
        _RetrySubstore_testResultIngestIndexById.set(this, new Map());
    }
    recordIngestOrder(testResultId) {
        if (__classPrivateFieldGet(this, _RetrySubstore_testResultIngestIndexById, "f").has(testResultId)) {
            return;
        }
        __classPrivateFieldGet(this, _RetrySubstore_testResultIngestIndexById, "f").set(testResultId, __classPrivateFieldGet(this, _RetrySubstore_testResultIngestIndexById, "f").size);
    }
    restoreIngestOrder(restoredIds, hasTestResult) {
        __classPrivateFieldGet(this, _RetrySubstore_testResultIngestIndexById, "f").clear();
        for (const id of restoredIds ?? []) {
            if (!hasTestResult(id) || __classPrivateFieldGet(this, _RetrySubstore_testResultIngestIndexById, "f").has(id)) {
                continue;
            }
            __classPrivateFieldGet(this, _RetrySubstore_testResultIngestIndexById, "f").set(id, __classPrivateFieldGet(this, _RetrySubstore_testResultIngestIndexById, "f").size);
        }
    }
    ingestOrderIdsForDump() {
        return Array.from(__classPrivateFieldGet(this, _RetrySubstore_testResultIngestIndexById, "f").keys());
    }
    upsert(testResult) {
        testResult.isRetry = false;
        if (!testResult.retryHash) {
            return;
        }
        const results = __classPrivateFieldGet(this, _RetrySubstore_testResultsByRetryHash, "f").get(testResult.retryHash);
        if (!results) {
            __classPrivateFieldGet(this, _RetrySubstore_testResultsByRetryHash, "f").set(testResult.retryHash, [testResult]);
            return;
        }
        results.push(testResult);
        results.sort((first, second) => __classPrivateFieldGet(this, _RetrySubstore_instances, "m", _RetrySubstore_compareResults).call(this, first, second));
        results.forEach((attempt, index) => {
            attempt.isRetry = index !== 0;
        });
    }
    retriesByTr(testResult) {
        if (!testResult.retryHash || testResult.isRetry) {
            return NO_RETRIES;
        }
        const attempts = __classPrivateFieldGet(this, _RetrySubstore_testResultsByRetryHash, "f").get(testResult.retryHash) ?? [];
        const index = attempts.findIndex((attempt) => attempt.id === testResult.id);
        if (index !== 0) {
            return NO_RETRIES;
        }
        return attempts.slice(1);
    }
    reset() {
        __classPrivateFieldGet(this, _RetrySubstore_testResultsByRetryHash, "f").clear();
    }
}
_RetrySubstore_testResultsByRetryHash = new WeakMap(), _RetrySubstore_testResultIngestIndexById = new WeakMap(), _RetrySubstore_instances = new WeakSet(), _RetrySubstore_getIngestIndex = function _RetrySubstore_getIngestIndex(testResultId) {
    return __classPrivateFieldGet(this, _RetrySubstore_testResultIngestIndexById, "f").get(testResultId) ?? -1;
}, _RetrySubstore_compareResults = function _RetrySubstore_compareResults(first, second) {
    const firstHasStart = typeof first.start === "number";
    const secondHasStart = typeof second.start === "number";
    if (firstHasStart !== secondHasStart) {
        return firstHasStart ? 1 : -1;
    }
    if (firstHasStart && secondHasStart && first.start !== second.start) {
        return first.start > second.start ? -1 : 1;
    }
    const indexA = __classPrivateFieldGet(this, _RetrySubstore_instances, "m", _RetrySubstore_getIngestIndex).call(this, first.id);
    const indexB = __classPrivateFieldGet(this, _RetrySubstore_instances, "m", _RetrySubstore_getIngestIndex).call(this, second.id);
    if (indexA !== indexB) {
        return indexA > indexB ? -1 : 1;
    }
    return 0;
};
