// @allurereport/plugin-testops/dist/gitFlow/LaunchGitFlow.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _LaunchGitFlow_instances, _LaunchGitFlow_ci, _LaunchGitFlow_gitFlow, _LaunchGitFlow_ancestorLimit, _LaunchGitFlow_logger, _LaunchGitFlow_classify;
import { collectGitFacts, isGitAvailable } from "@allurereport/git";
import { buildGitFlowContext, shouldAttachGitFlow } from "./context.js";
import { projectLaunchGitContext } from "./projection.js";
export class LaunchGitFlow {
    constructor(params) {
        _LaunchGitFlow_instances.add(this);
        _LaunchGitFlow_ci.set(this, void 0);
        _LaunchGitFlow_gitFlow.set(this, void 0);
        _LaunchGitFlow_ancestorLimit.set(this, void 0);
        _LaunchGitFlow_logger.set(this, void 0);
        __classPrivateFieldSet(this, _LaunchGitFlow_ci, params.ci, "f");
        __classPrivateFieldSet(this, _LaunchGitFlow_gitFlow, params.gitFlow, "f");
        __classPrivateFieldSet(this, _LaunchGitFlow_ancestorLimit, params.ancestorLimit, "f");
        __classPrivateFieldSet(this, _LaunchGitFlow_logger, params.logger, "f");
    }
    resolve() {
        if (!shouldAttachGitFlow(__classPrivateFieldGet(this, _LaunchGitFlow_gitFlow, "f"))) {
            return undefined;
        }
        if (!isGitAvailable()) {
            __classPrivateFieldGet(this, _LaunchGitFlow_logger, "f").warn("git CLI is not available; continuing upload without git context");
            return undefined;
        }
        const facts = collectGitFacts({ ancestorLimit: __classPrivateFieldGet(this, _LaunchGitFlow_ancestorLimit, "f") });
        const gitFlowContext = buildGitFlowContext({
            ci: __classPrivateFieldGet(this, _LaunchGitFlow_ci, "f"),
            facts,
            ancestorLimit: __classPrivateFieldGet(this, _LaunchGitFlow_ancestorLimit, "f"),
        });
        if (!gitFlowContext) {
            __classPrivateFieldGet(this, _LaunchGitFlow_logger, "f").warn("Git Flow metadata could not be collected; continuing upload without git context");
            return undefined;
        }
        const classified = __classPrivateFieldGet(this, _LaunchGitFlow_instances, "m", _LaunchGitFlow_classify).call(this, gitFlowContext);
        if (!classified) {
            __classPrivateFieldGet(this, _LaunchGitFlow_logger, "f").warn("Pull request git context is incomplete (missing source or target branch); continuing upload without git context");
            return undefined;
        }
        const launchGitContext = projectLaunchGitContext(classified, __classPrivateFieldGet(this, _LaunchGitFlow_ci, "f"));
        if (!launchGitContext) {
            __classPrivateFieldGet(this, _LaunchGitFlow_logger, "f").warn("Git provider is not supported by TestOps; continuing upload without git context");
            return undefined;
        }
        return launchGitContext;
    }
}
_LaunchGitFlow_ci = new WeakMap(), _LaunchGitFlow_gitFlow = new WeakMap(), _LaunchGitFlow_ancestorLimit = new WeakMap(), _LaunchGitFlow_logger = new WeakMap(), _LaunchGitFlow_instances = new WeakSet(), _LaunchGitFlow_classify = function _LaunchGitFlow_classify(context) {
    if (context.pullRequest) {
        const sourceBranch = context.branch;
        const targetBranch = context.targetBranch;
        if (!sourceBranch || !targetBranch) {
            return undefined;
        }
        return {
            kind: "pull_request",
            context,
            sourceBranch,
            targetBranch,
        };
    }
    if (context.branch) {
        return {
            kind: "branch",
            context,
            branch: context.branch,
        };
    }
    return {
        kind: "standalone",
        context,
    };
};
