// @allurereport/plugin-testops/dist/gitFlow/projection.js
import { GitProvider } from "@allurereport/core-api";
const mapProvider = (provider) => {
    switch (provider) {
        case GitProvider.Github:
            return "github";
        case GitProvider.Gitlab:
            return "gitlab";
        case GitProvider.Bitbucket:
            return "bitbucket";
        default:
            return undefined;
    }
};
const hostMatchesProvider = (host, providerType) => {
    switch (providerType) {
        case "github":
            return host === "github.com" || host.endsWith(".github.com") || host.startsWith("github.");
        case "gitlab":
            return host === "gitlab.com" || host.endsWith(".gitlab.com") || host.startsWith("gitlab.");
        case "bitbucket":
            return host === "bitbucket.org" || host.endsWith(".bitbucket.org") || host.startsWith("bitbucket.");
        default:
            return false;
    }
};
const inferProviderTypeFromUrl = (url) => {
    if (!url) {
        return undefined;
    }
    try {
        const host = new URL(url).hostname.toLowerCase();
        for (const providerType of ["github", "gitlab", "bitbucket"]) {
            if (hostMatchesProvider(host, providerType)) {
                return providerType;
            }
        }
    }
    catch {
        return undefined;
    }
    return undefined;
};
const resolveProviderType = (context) => mapProvider(context.provider) ?? inferProviderTypeFromUrl(context.repository.url);
const stripGitSuffix = (url) => url.replace(/\.git\/?$/i, "");
const buildRepositoryUrl = (providerType, slug, repositoryUrl) => {
    if (repositoryUrl) {
        return repositoryUrl;
    }
    if (providerType === "github") {
        return `https://github.com/${slug}.git`;
    }
    if (providerType === "gitlab") {
        return `https://gitlab.com/${slug}`;
    }
    if (providerType === "bitbucket") {
        return `https://bitbucket.org/${slug}.git`;
    }
    return undefined;
};
const buildRepositoryWebUrl = (providerType, slug, repositoryUrl) => {
    if (repositoryUrl) {
        return stripGitSuffix(repositoryUrl);
    }
    if (providerType === "github") {
        return `https://github.com/${slug}`;
    }
    if (providerType === "gitlab") {
        return `https://gitlab.com/${slug}`;
    }
    if (providerType === "bitbucket") {
        return `https://bitbucket.org/${slug}`;
    }
    return repositoryUrl;
};
const buildCommitUrl = (repoWebUrl, hash, providerType) => {
    if (!repoWebUrl) {
        return undefined;
    }
    if (providerType === "github") {
        return `${repoWebUrl}/commit/${hash}`;
    }
    if (providerType === "gitlab") {
        return `${repoWebUrl}/-/commit/${hash}`;
    }
    return `${repoWebUrl}/commits/${hash}`;
};
const buildBranchUrl = (repoWebUrl, branchName, providerType) => {
    if (!repoWebUrl) {
        return undefined;
    }
    if (providerType === "github") {
        return `${repoWebUrl}/tree/${branchName}`;
    }
    if (providerType === "gitlab") {
        return `${repoWebUrl}/-/tree/${branchName}`;
    }
    return `${repoWebUrl}/branch/${branchName}`;
};
const buildPullRequestUrl = (repoWebUrl, externalId, providerType, pullRequestUrl) => {
    if (pullRequestUrl) {
        return pullRequestUrl;
    }
    if (!repoWebUrl) {
        return undefined;
    }
    if (providerType === "github") {
        return `${repoWebUrl}/pull/${externalId}`;
    }
    if (providerType === "gitlab") {
        return `${repoWebUrl}/-/merge_requests/${externalId}`;
    }
    return `${repoWebUrl}/pull-requests/${externalId}`;
};
const toStandaloneMetadata = (localState) => {
    if (!localState) {
        return undefined;
    }
    return {
        hasUncommittedChanges: localState.uncommittedChanges,
        isUnpublishedCommit: localState.unpublishedCommit,
        isUnpublishedBranch: localState.unpublishedBranch,
    };
};
const toPullRequestMetadata = (ci) => {
    if (!ci?.jobRunUid && !ci?.jobRunName) {
        return undefined;
    }
    return {
        workflowRunId: ci.jobRunUid || undefined,
        workflowRunName: ci.jobRunName || undefined,
    };
};
const toBranchDto = (name, providerType, repoWebUrl) => ({
    name,
    url: buildBranchUrl(repoWebUrl, name, providerType),
});
const buildRepositoryAndCommit = (context, providerType) => {
    const repositoryUrl = buildRepositoryUrl(providerType, context.repository.slug, context.repository.url);
    const repoWebUrl = buildRepositoryWebUrl(providerType, context.repository.slug, repositoryUrl);
    const lineage = context.firstParentAncestors.length > 0 ? context.firstParentAncestors : undefined;
    return {
        repository: {
            providerType,
            name: context.repository.slug,
            url: repositoryUrl,
        },
        commit: {
            hash: context.commit,
            url: buildCommitUrl(repoWebUrl, context.commit, providerType),
            lineage,
        },
        repoWebUrl,
    };
};
export const projectLaunchGitContext = (classified, ci) => {
    const context = classified.context;
    const providerType = resolveProviderType(context);
    if (!providerType) {
        return undefined;
    }
    const { repository, commit, repoWebUrl } = buildRepositoryAndCommit(context, providerType);
    if (classified.kind === "pull_request") {
        const { pullRequest } = context;
        if (!pullRequest) {
            return undefined;
        }
        return {
            contextType: "pull_request",
            repository,
            commit,
            pullRequest: {
                externalId: pullRequest.id,
                title: pullRequest.title,
                url: buildPullRequestUrl(repoWebUrl, pullRequest.id, providerType, pullRequest.url),
                sourceBranch: toBranchDto(classified.sourceBranch, providerType, repoWebUrl),
                targetBranch: toBranchDto(classified.targetBranch, providerType, repoWebUrl),
            },
            metadata: toPullRequestMetadata(ci),
        };
    }
    if (classified.kind === "branch") {
        return {
            contextType: "branch",
            repository,
            commit,
            branch: toBranchDto(classified.branch, providerType, repoWebUrl),
        };
    }
    return {
        contextType: "standalone",
        repository,
        commit,
        metadata: toStandaloneMetadata(context.localState),
    };
};
