// allure/dist/utils/process.js
import { spawn } from "node:child_process";
import process from "node:process";
import { platform } from "process";
import { invokeStdoutCliTool } from "@allurereport/reader-api";
const IS_WIN = platform === "win32";
const PS_OUTPUT_PATTERN = /(?<ppid>\d+)\s+(?<pid>\d+)\s+(?<comm>.*)/;
const quotePosixShellArg = (arg) => {
    if (/^[A-Za-z0-9_/:=.,@%+-]+$/u.test(arg)) {
        return arg;
    }
    return `'${arg.replace(/'/g, "'\\''")}'`;
};
const quoteWindowsShellArg = (arg) => {
    if (arg && !/[\s"&()<>^|%]/u.test(arg)) {
        return arg;
    }
    let quoted = '"';
    let backslashes = 0;
    for (const char of arg) {
        if (char === "\\") {
            backslashes += 1;
            continue;
        }
        if (char === '"') {
            quoted += "\\".repeat(backslashes * 2 + 1);
            quoted += char;
            backslashes = 0;
            continue;
        }
        quoted += "\\".repeat(backslashes);
        backslashes = 0;
        quoted += char;
    }
    quoted += "\\".repeat(backslashes * 2);
    quoted += '"';
    return quoted;
};
const quoteShellArg = IS_WIN ? quoteWindowsShellArg : quotePosixShellArg;
const resolveShellInvocation = (command, commandArgs, shell) => {
    if (!shell || commandArgs.length === 0) {
        return {
            command,
            commandArgs,
        };
    }
    return {
        command: [command, ...commandArgs].map(quoteShellArg).join(" "),
        commandArgs: [],
    };
};
export const runProcess = (params) => {
    const { command, commandArgs, cwd, environmentVariables = {}, logs = "inherit", shell = IS_WIN } = params;
    const env = {
        ...process.env,
        ...environmentVariables,
    };
    if (logs === "pipe") {
        Object.assign(env, {
            FORCE_COLOR: "1",
            CLICOLOR_FORCE: "1",
            COLOR: "1",
            COLORTERM: "truecolor",
            TERM: "xterm-256color",
        });
        delete env.NO_COLOR;
    }
    const processInvocation = resolveShellInvocation(command, commandArgs, shell);
    return spawn(processInvocation.command, processInvocation.commandArgs, {
        env,
        cwd,
        stdio: logs,
        shell,
    });
};
export const terminationOf = (testProcess) => new Promise((resolve) => {
    testProcess.on("exit", (code) => resolve(code));
});
export const stopProcessTree = async (pid, { signal = "SIGTERM" } = {}) => {
    const tree = new Map();
    const processesToSignal = [];
    const signaledProcesses = [];
    const [executable, processArgs] = IS_WIN
        ? [
            "powershell.exe",
            [
                "-NoLogo",
                "-NoProfile",
                "-NonInteractive",
                "-Command",
                `& {
            [System.Console]::OutputEncoding = [System.Text.UTF8Encoding]::new($False, $False)
            Get-CimInstance -Class Win32_Process | ForEach-Object {
              "$($_.ParentProcessId) $($_.ProcessId) $($_.ExecutablePath)"
            }
          }`,
            ],
        ]
        : ["ps", ["-Ao", "ppid,pid,comm"]];
    const processOptions = {
        timeout: 5000,
        encoding: "utf-8",
    };
    for await (const line of invokeStdoutCliTool(executable, processArgs, processOptions)) {
        const match = PS_OUTPUT_PATTERN.exec(line);
        if (match) {
            const [, ppidStr, pidStr, comm] = match;
            const ppid = parseInt(ppidStr, 10);
            const entryPid = parseInt(pidStr, 10);
            const processBrief = {
                pid: entryPid,
                parentPid: ppid,
                command: comm,
            };
            if (entryPid === pid) {
                processesToSignal.push(processBrief);
            }
            const children = tree.get(ppid) ?? [];
            if (!children.length) {
                tree.set(ppid, children);
            }
            children.push(processBrief);
        }
    }
    const pidStack = [];
    for (let currentPid = pid; currentPid; currentPid = pidStack.shift()) {
        const children = tree.get(currentPid);
        if (children) {
            processesToSignal.push(...children);
            pidStack.push(...children.map(({ pid: childPid }) => childPid));
        }
    }
    for (const target of processesToSignal) {
        try {
            process.kill(target.pid, signal);
            signaledProcesses.push(target);
        }
        catch {
        }
    }
    return signaledProcesses;
};
