// @allurereport/ci/dist/detectors/azure.js
import { CiType, GitProvider } from "@allurereport/core-api";
import { resolveRepositoryFromGitUrl } from "../helpers/gitProvider.js";
import { getEnv } from "../utils.js";
const REF_PREFIX = "refs/";
const BRANCH_REF_PREFIX = "refs/heads/";
const TAG_REF_PREFIX = "refs/tags/";
export const getRootURL = () => getEnv("SYSTEM_COLLECTIONURI");
export const getBuildID = () => getEnv("BUILD_BUILDID");
export const getDefinitionID = () => getEnv("SYSTEM_DEFINITIONID");
export const getProjectID = () => getEnv("SYSTEM_TEAMPROJECTID");
const mapAzureRepositoryProvider = (provider) => {
    switch (provider) {
        case "GitHub":
            return GitProvider.Github;
        case "Bitbucket":
            return GitProvider.Bitbucket;
        default:
            return undefined;
    }
};
const normalizeBranchRef = (branch) => {
    const trimmed = branch?.trim() ?? "";
    if (!trimmed || trimmed.startsWith(TAG_REF_PREFIX)) {
        return undefined;
    }
    if (trimmed.startsWith(BRANCH_REF_PREFIX)) {
        const branchName = trimmed.slice(BRANCH_REF_PREFIX.length);
        return branchName || undefined;
    }
    if (trimmed.startsWith(REF_PREFIX)) {
        return undefined;
    }
    return trimmed;
};
const getSourceBranch = () => {
    const sourceBranch = getEnv("BUILD_SOURCEBRANCH");
    if (sourceBranch) {
        return normalizeBranchRef(sourceBranch);
    }
    return normalizeBranchRef(getEnv("BUILD_SOURCEBRANCHNAME"));
};
const getPullRequestId = () => {
    return getEnv("SYSTEM_PULLREQUEST_PULLREQUESTNUMBER") || getEnv("SYSTEM_PULLREQUEST_PULLREQUESTID") || undefined;
};
const getRepositoryUrl = () => getEnv("BUILD_REPOSITORY_URI") || getEnv("SYSTEM_PULLREQUEST_SOURCEREPOSITORYURI");
const getRepositoryFromUrl = () => {
    const repositoryUrl = getRepositoryUrl();
    return repositoryUrl ? resolveRepositoryFromGitUrl(repositoryUrl) : undefined;
};
export const azure = {
    type: CiType.Azure,
    get detected() {
        return getEnv("SYSTEM_DEFINITIONID") !== "";
    },
    get repoName() {
        const repoName = getEnv("BUILD_REPOSITORY_NAME");
        return repoName.split("/")?.[1] ?? repoName;
    },
    get jobUid() {
        return `${getProjectID()}_${getDefinitionID()}`;
    },
    get jobUrl() {
        return `${getRootURL()}/${getProjectID()}/_build?definitionId=${getDefinitionID()}`;
    },
    get jobName() {
        return getEnv("BUILD_DEFINITIONNAME");
    },
    get jobRunUid() {
        return getBuildID();
    },
    get jobRunUrl() {
        return `${getRootURL()}/${getProjectID()}/_build/results?buildId=${getBuildID()}`;
    },
    get jobRunName() {
        return getEnv("BUILD_BUILDNUMBER");
    },
    get jobRunBranch() {
        return getSourceBranch() ?? "";
    },
    get pullRequestUrl() {
        const repositoryProvider = getEnv("BUILD_REPOSITORY_PROVIDER");
        const repositoryUrl = getEnv("SYSTEM_PULLREQUEST_SOURCEREPOSITORYURI") || getEnv("BUILD_REPOSITORY_URI");
        const pullRequestId = getPullRequestId();
        if (!repositoryUrl || !pullRequestId) {
            return "";
        }
        if (repositoryProvider === "GitHub") {
            return `${repositoryUrl}/pull/${pullRequestId}`;
        }
        if (repositoryProvider === "TfsGit" || repositoryProvider === "TfsVersionControl") {
            return `${repositoryUrl}/pullrequest/${pullRequestId}`;
        }
        return "";
    },
    get pullRequestName() {
        return "";
    },
    get provider() {
        return getRepositoryFromUrl()?.provider ?? mapAzureRepositoryProvider(getEnv("BUILD_REPOSITORY_PROVIDER"));
    },
    get repository() {
        const repositoryUrl = getRepositoryUrl();
        const fromUrl = repositoryUrl ? resolveRepositoryFromGitUrl(repositoryUrl) : undefined;
        const slug = fromUrl?.slug ?? (getEnv("BUILD_REPOSITORY_NAME") || undefined);
        if (!this.provider || !slug) {
            return undefined;
        }
        return {
            slug,
            url: fromUrl?.url ?? repositoryUrl,
        };
    },
    get sourceBranch() {
        return normalizeBranchRef(getEnv("SYSTEM_PULLREQUEST_SOURCEBRANCH")) || getSourceBranch() || undefined;
    },
    get targetBranch() {
        return (normalizeBranchRef(getEnv("SYSTEM_PULLREQUEST_TARGETBRANCH")) ||
            normalizeBranchRef(getEnv("SYSTEM_PULLREQUEST_TARGETBRANCHNAME")) ||
            undefined);
    },
    get pullRequest() {
        const pullRequestNumber = getPullRequestId();
        return pullRequestNumber
            ? {
                id: pullRequestNumber,
                url: this.pullRequestUrl || undefined,
                title: this.pullRequestName || undefined,
            }
            : undefined;
    },
};
