// @allurereport/ci/dist/detectors/bitbucket.js
import { CiType, GitProvider } from "@allurereport/core-api";
import { getEnv } from "../utils.js";
const stripGitSuffix = (url) => url.replace(/\.git\/?$/i, "");
const getRepositoryUrl = () => {
    const origin = (getEnv("BITBUCKET_GIT_HTTP_ORIGIN") || "").trim();
    if (!origin) {
        return "";
    }
    try {
        const url = new URL(origin);
        url.username = "";
        url.password = "";
        url.pathname = stripGitSuffix(url.pathname).replace(/\/+$/, "");
        url.search = "";
        url.hash = "";
        return url.toString().replace(/\/$/, "");
    }
    catch {
        return stripGitSuffix(origin).replace(/\/+$/, "");
    }
};
export const getJobURL = () => {
    const origin = getRepositoryUrl();
    return `${origin}/pipelines`;
};
export const bitbucket = {
    type: CiType.Bitbucket,
    get detected() {
        return getEnv("BITBUCKET_PIPELINE_UUID") !== "";
    },
    get repoName() {
        return getEnv("BITBUCKET_REPO_SLUG");
    },
    get jobUid() {
        return getEnv("BITBUCKET_REPO_FULL_NAME");
    },
    get jobUrl() {
        return getJobURL();
    },
    get jobName() {
        return getEnv("BITBUCKET_REPO_FULL_NAME");
    },
    get jobRunUid() {
        return getEnv("BITBUCKET_PIPELINE_UUID");
    },
    get jobRunUrl() {
        return `${getJobURL()}/results/${this.jobRunUid}`;
    },
    get jobRunName() {
        return getEnv("BITBUCKET_PIPELINE_UUID");
    },
    get jobRunBranch() {
        return getEnv("BITBUCKET_BRANCH");
    },
    get pullRequestUrl() {
        const prId = getEnv("BITBUCKET_PR_ID");
        if (!prId) {
            return "";
        }
        const origin = getRepositoryUrl();
        if (!origin) {
            return "";
        }
        return `${origin}/pull-requests/${prId}`;
    },
    get pullRequestName() {
        return "";
    },
    get provider() {
        return GitProvider.Bitbucket;
    },
    get repository() {
        const repositorySlug = getEnv("BITBUCKET_REPO_FULL_NAME");
        return repositorySlug
            ? {
                slug: repositorySlug,
                url: getRepositoryUrl() || undefined,
            }
            : undefined;
    },
    get sourceBranch() {
        return getEnv("BITBUCKET_BRANCH") || this.jobRunBranch || undefined;
    },
    get targetBranch() {
        return getEnv("BITBUCKET_PR_DESTINATION_BRANCH") || undefined;
    },
    get pullRequest() {
        const prId = getEnv("BITBUCKET_PR_ID");
        return prId
            ? {
                id: prId,
                url: this.pullRequestUrl || undefined,
                title: this.pullRequestName || undefined,
            }
            : undefined;
    },
};
