// @allurereport/ci/dist/detectors/circle.js
import { CiType } from "@allurereport/core-api";
import { parsePullRequestNumberFromUrl, resolveRepositoryFromGitUrl } from "../helpers/gitProvider.js";
import { getEnv, parseURLPath } from "../utils.js";
export const getBuildNumber = () => getEnv("CIRCLE_BUILD_NUM");
export const getJobRunURL = () => getEnv("CIRCLE_BUILD_URL");
const getJobURL = () => {
    const jobRunURL = getJobRunURL();
    const buildNumber = getBuildNumber();
    return jobRunURL.replace(`/${buildNumber}`, "");
};
const getRepository = () => {
    const repositoryUrl = getEnv("CIRCLE_REPOSITORY_URL");
    return repositoryUrl ? resolveRepositoryFromGitUrl(repositoryUrl) : undefined;
};
const getPullRequestUrl = () => {
    const pullRequestUrl = getEnv("CIRCLE_PULL_REQUEST").trim();
    if (pullRequestUrl) {
        return pullRequestUrl;
    }
    return (getEnv("CIRCLE_PULL_REQUESTS")
        .split(",")
        .map((url) => url.trim())
        .find(Boolean) ?? "");
};
const getPullRequestNumber = () => {
    const pullRequestUrl = getPullRequestUrl();
    const pullRequestNumber = pullRequestUrl ? parsePullRequestNumberFromUrl(pullRequestUrl) : undefined;
    return pullRequestNumber || getEnv("CIRCLE_PR_NUMBER") || undefined;
};
export const circle = {
    type: CiType.Circle,
    get detected() {
        const hasEnv = getEnv("CIRCLECI") !== "";
        const jobURL = getJobURL();
        const path = parseURLPath(jobURL);
        return hasEnv && path !== "";
    },
    get repoName() {
        return getEnv("CIRCLE_PROJECT_REPONAME");
    },
    get jobUid() {
        const jobURL = getJobURL();
        return parseURLPath(jobURL);
    },
    get jobUrl() {
        return getJobURL();
    },
    get jobName() {
        const username = getEnv("CIRCLE_PROJECT_USERNAME") || getEnv("CIRCLE_USERNAME");
        const reponame = getEnv("CIRCLE_PROJECT_REPONAME");
        return `${username}/${reponame}`;
    },
    get jobRunUid() {
        return getEnv("CIRCLE_WORKFLOW_JOB_ID");
    },
    get jobRunUrl() {
        return getJobRunURL();
    },
    get jobRunName() {
        return getEnv("CIRCLE_BUILD_NUM");
    },
    get jobRunBranch() {
        return getEnv("CIRCLE_BRANCH");
    },
    get pullRequestUrl() {
        return getPullRequestUrl();
    },
    get pullRequestName() {
        return "";
    },
    get provider() {
        return getRepository()?.provider;
    },
    get repository() {
        const repository = getRepository();
        return repository
            ? {
                slug: repository.slug,
                url: repository.url,
            }
            : undefined;
    },
    get sourceBranch() {
        return getEnv("CIRCLE_BRANCH") || this.jobRunBranch || undefined;
    },
    get pullRequest() {
        const pullRequestUrl = getPullRequestUrl();
        const pullRequestNumber = getPullRequestNumber();
        return pullRequestNumber
            ? {
                id: pullRequestNumber,
                url: this.pullRequestUrl || pullRequestUrl || undefined,
                title: this.pullRequestName || undefined,
            }
            : undefined;
    },
};
