// @allurereport/ci/dist/detectors/local.js
import { spawnSync } from "node:child_process";
import { basename } from "node:path";
import { CiType } from "@allurereport/core-api";
import { resolveRepositoryFromGitUrl } from "../helpers/gitProvider.js";
const readGitRemoteUrl = () => {
    const output = spawnSync("git", ["remote", "get-url", "origin"]);
    if (output.error) {
        return "";
    }
    return output.stdout.toString().trim();
};
const getRepository = () => {
    const remoteUrl = readGitRemoteUrl();
    return remoteUrl ? resolveRepositoryFromGitUrl(remoteUrl) : undefined;
};
export const local = {
    type: CiType.Local,
    get detected() {
        return true;
    },
    get repoName() {
        const output = spawnSync("git", ["rev-parse", "--show-toplevel"]);
        if (output.error) {
            return "";
        }
        return basename(output.stdout.toString().trim());
    },
    get jobUid() {
        return "";
    },
    get jobUrl() {
        return "";
    },
    get jobName() {
        return "";
    },
    get jobRunUid() {
        return "";
    },
    get jobRunUrl() {
        return "";
    },
    get jobRunName() {
        return "";
    },
    get jobRunBranch() {
        const output = spawnSync("git", ["rev-parse", "--abbrev-ref", "HEAD"]);
        if (output.error) {
            return "";
        }
        return output.stdout.toString().trim();
    },
    get pullRequestUrl() {
        return "";
    },
    get pullRequestName() {
        return "";
    },
    get provider() {
        return getRepository()?.provider;
    },
    get repository() {
        const repository = getRepository();
        return repository
            ? {
                slug: repository.slug,
                url: repository.url,
            }
            : undefined;
    },
    get sourceBranch() {
        return this.jobRunBranch || undefined;
    },
};
export const isLocalCiDescriptor = (ci) => ci.type === CiType.Local;
