// @allurereport/plugin-agent/dist/state.js
import { appendFile, readFile, rm } from "node:fs/promises";
import { join, resolve } from "node:path";
import { isFileNotFoundError, listAgentManagedTempOutputDirs, listAgentStatePaths, pathExists, projectStatePath, readPathMtimeMs, tryWithAgentStateLock, withAgentStateLock, writeJsonlAtomic, } from "./utils.js";
export { ALLURE_AGENT_STATE_DIR_ENV, resolveAgentStateDir } from "./utils.js";
const AGENT_RUN_STATE_SCHEMA = "allure-agent-run/v1";
const AGENT_STALE_OUTPUT_TTL_MS = 7 * 24 * 60 * 60 * 1000;
const isAgentRunState = (value) => {
    if (typeof value !== "object" || value === null) {
        return false;
    }
    const candidate = value;
    return (candidate.schema === AGENT_RUN_STATE_SCHEMA &&
        typeof candidate.runId === "string" &&
        typeof candidate.cwd === "string" &&
        typeof candidate.outputDir === "string" &&
        typeof candidate.managedOutput === "boolean" &&
        typeof candidate.command === "string" &&
        typeof candidate.startedAt === "number" &&
        Number.isSafeInteger(candidate.startedAt) &&
        (candidate.expectationsPath === undefined || typeof candidate.expectationsPath === "string") &&
        (candidate.finishedAt === undefined ||
            (typeof candidate.finishedAt === "number" && Number.isSafeInteger(candidate.finishedAt))) &&
        (candidate.status === "running" || candidate.status === "finished") &&
        (candidate.exitCode === undefined || typeof candidate.exitCode === "number" || candidate.exitCode === null) &&
        (candidate.pid === undefined || (typeof candidate.pid === "number" && Number.isSafeInteger(candidate.pid))));
};
const normalizeAgentRunState = (value) => ({
    schema: AGENT_RUN_STATE_SCHEMA,
    runId: value.runId,
    cwd: resolve(value.cwd),
    outputDir: resolve(value.outputDir),
    managedOutput: value.managedOutput,
    expectationsPath: value.expectationsPath ? resolve(value.expectationsPath) : undefined,
    command: value.command,
    startedAt: value.startedAt,
    finishedAt: value.finishedAt,
    status: value.status,
    exitCode: value.exitCode,
    pid: value.pid,
});
const readAgentRunStateFile = async (statePath, cwd) => {
    const normalizedCwd = cwd === undefined ? undefined : resolve(cwd);
    let raw;
    try {
        raw = await readFile(statePath, "utf-8");
    }
    catch (error) {
        if (isFileNotFoundError(error)) {
            return [];
        }
        throw error;
    }
    return raw
        .split("\n")
        .map((line) => line.trim())
        .filter(Boolean)
        .flatMap((line) => {
        try {
            const parsed = JSON.parse(line);
            if (!isAgentRunState(parsed)) {
                return [];
            }
            return normalizedCwd === undefined || parsed.cwd === normalizedCwd ? [parsed] : [];
        }
        catch {
            return [];
        }
    });
};
const readAgentRunStateLines = async (cwd) => {
    const normalizedCwd = resolve(cwd);
    return readAgentRunStateFile(projectStatePath(normalizedCwd), normalizedCwd);
};
const foldAgentRunStates = (states) => {
    const order = [];
    const latestByRunId = new Map();
    for (const state of states) {
        if (!latestByRunId.has(state.runId)) {
            order.push(state.runId);
        }
        latestByRunId.set(state.runId, state);
    }
    return order
        .map((runId) => latestByRunId.get(runId))
        .filter((state) => state !== undefined)
        .sort((a, b) => a.startedAt - b.startedAt || (a.finishedAt ?? a.startedAt) - (b.finishedAt ?? b.startedAt));
};
const getAgentRunStateAgeTimestamp = (state) => state.finishedAt ?? state.startedAt;
const isManagedOutputStale = (state, now, staleOutputTtlMs) => state.managedOutput && now - getAgentRunStateAgeTimestamp(state) >= staleOutputTtlMs;
const isAgentOutputDirectory = async (outputDir) => (await pathExists(join(outputDir, "manifest", "run.json"))) || (await pathExists(join(outputDir, "index.md")));
const cleanupStaleAgentRunState = async (params) => {
    const states = foldAgentRunStates(await readAgentRunStateLines(params.cwd));
    const retained = [];
    const deleted = [];
    const failed = [];
    for (const state of states) {
        if (!(await pathExists(state.outputDir))) {
            continue;
        }
        if (!isManagedOutputStale(state, params.now, params.staleOutputTtlMs)) {
            retained.push(state);
            continue;
        }
        try {
            await rm(state.outputDir, { recursive: true, force: true });
            deleted.push(state);
        }
        catch (error) {
            retained.push(state);
            failed.push({ state, error });
        }
    }
    await writeJsonlAtomic(projectStatePath(params.cwd), retained);
    return {
        deleted,
        failed,
        retained,
    };
};
const cleanupStaleOrphanAgentOutputs = async (params) => {
    const outputDirs = await listAgentManagedTempOutputDirs();
    const deleted = [];
    const failed = [];
    const retained = [];
    for (const outputDir of outputDirs) {
        const normalizedOutputDir = resolve(outputDir);
        if (params.referencedOutputDirs.has(normalizedOutputDir)) {
            retained.push(normalizedOutputDir);
            continue;
        }
        let mtimeMs;
        try {
            mtimeMs = await readPathMtimeMs(normalizedOutputDir);
        }
        catch (error) {
            if (!isFileNotFoundError(error)) {
                failed.push({ outputDir: normalizedOutputDir, error });
            }
            continue;
        }
        if (params.now - mtimeMs < params.staleOutputTtlMs) {
            retained.push(normalizedOutputDir);
            continue;
        }
        if (!(await isAgentOutputDirectory(normalizedOutputDir))) {
            retained.push(normalizedOutputDir);
            continue;
        }
        try {
            await rm(normalizedOutputDir, { recursive: true, force: true });
            deleted.push(normalizedOutputDir);
        }
        catch (error) {
            failed.push({ outputDir: normalizedOutputDir, error });
        }
    }
    return {
        deleted,
        failed,
        retained,
    };
};
export const writeAgentRunState = async (value) => {
    const normalizedState = normalizeAgentRunState(value);
    await withAgentStateLock(normalizedState.cwd, async () => {
        await appendFile(projectStatePath(normalizedState.cwd), `${JSON.stringify(normalizedState)}\n`, "utf-8");
    });
    return normalizedState;
};
export const writeLatestAgentState = writeAgentRunState;
export const readAgentRunStates = async (cwd) => foldAgentRunStates(await readAgentRunStateLines(cwd));
export const readLatestAgentState = async (cwd) => {
    const states = await readAgentRunStates(cwd);
    for (let i = states.length - 1; i >= 0; i -= 1) {
        const state = states[i];
        if (await pathExists(state.outputDir)) {
            return state;
        }
    }
    return undefined;
};
export const cleanupAgentRunState = async (params) => withAgentStateLock(params.cwd, async () => {
    const keepManagedRuns = Math.max(0, params.keepManagedRuns ?? 1);
    const states = foldAgentRunStates(await readAgentRunStateLines(params.cwd));
    const existing = [];
    for (const state of states) {
        if (await pathExists(state.outputDir)) {
            existing.push(state);
        }
    }
    const retainedManagedRunIds = new Set(existing
        .filter((state) => state.managedOutput && state.status === "finished")
        .sort((a, b) => (b.finishedAt ?? b.startedAt) - (a.finishedAt ?? a.startedAt) || b.startedAt - a.startedAt)
        .slice(0, keepManagedRuns)
        .map((state) => state.runId));
    if (params.currentRunId) {
        retainedManagedRunIds.add(params.currentRunId);
    }
    const deleted = [];
    const failed = [];
    for (const state of existing) {
        if (!state.managedOutput || state.status !== "finished" || retainedManagedRunIds.has(state.runId)) {
            continue;
        }
        try {
            await rm(state.outputDir, { recursive: true, force: true });
            deleted.push(state);
        }
        catch (error) {
            failed.push({ state, error });
        }
    }
    const deletedRunIds = new Set(deleted.map((state) => state.runId));
    const retained = existing.filter((state) => !deletedRunIds.has(state.runId));
    await writeJsonlAtomic(projectStatePath(params.cwd), retained);
    return {
        deleted,
        failed,
        retained,
    };
});
export const cleanupStaleAgentRunStates = async (params) => {
    const currentCwd = resolve(params.cwd);
    const currentStatePath = projectStatePath(currentCwd);
    const statePaths = await listAgentStatePaths(currentCwd);
    const staleOutputTtlMs = Math.max(0, params.staleOutputTtlMs ?? AGENT_STALE_OUTPUT_TTL_MS);
    const now = params.now ?? Date.now();
    const staleCwds = new Map();
    const referencedOutputDirs = new Set();
    for (const statePath of statePaths) {
        const states = foldAgentRunStates(await readAgentRunStateFile(statePath));
        for (const state of states) {
            referencedOutputDirs.add(resolve(state.outputDir));
            if (statePath === currentStatePath) {
                continue;
            }
            if (state.cwd === currentCwd || state.runId === params.currentRunId) {
                continue;
            }
            if (!(await pathExists(state.outputDir)) || isManagedOutputStale(state, now, staleOutputTtlMs)) {
                staleCwds.set(state.cwd, statePath);
                break;
            }
        }
    }
    const deleted = [];
    const failed = [];
    const retained = [];
    const skipped = [];
    for (const [cwd, statePath] of staleCwds) {
        const lockResult = await tryWithAgentStateLock(cwd, () => cleanupStaleAgentRunState({
            cwd,
            now,
            staleOutputTtlMs,
        }));
        if (!lockResult.acquired) {
            skipped.push({ cwd, statePath, reason: "locked" });
            continue;
        }
        deleted.push(...lockResult.result.deleted);
        failed.push(...lockResult.result.failed);
        retained.push(...lockResult.result.retained);
    }
    const orphaned = await cleanupStaleOrphanAgentOutputs({
        referencedOutputDirs,
        now,
        staleOutputTtlMs,
    });
    return {
        checked: statePaths.length,
        deleted,
        failed,
        orphaned,
        retained,
        skipped,
    };
};
