// @allurereport/service/dist/utils/http.js
import axios, { isAxiosError } from "axios";
import { isReportDataFile } from "./files.js";
export class KnownError extends Error {
    constructor(message, status) {
        super(message);
        this.name = "KnownError";
        this.status = status;
    }
}
export class UnknownError extends Error {
    constructor(message, stack) {
        super(message);
        this.name = "UnknownError";
        this.stack = stack;
    }
}
const ERROR_MESSAGE_FIELDS = ["message", "error_description", "error", "detail", "title", "description"];
const stringifyErrorObject = (value) => {
    try {
        return JSON.stringify(value);
    }
    catch {
        const entries = Object.entries(value).map(([key, entryValue]) => `${key}=${String(entryValue)}`);
        return entries.length > 0 ? `{ ${entries.join(", ")} }` : undefined;
    }
};
export const formatResponseErrorData = (data) => {
    if (data === undefined || data === null || data === "") {
        return undefined;
    }
    if (typeof data === "string") {
        return data;
    }
    if (typeof data === "number" || typeof data === "boolean" || typeof data === "bigint") {
        return String(data);
    }
    if (Array.isArray(data)) {
        const items = data.map(formatResponseErrorData).filter(Boolean);
        return items.length > 0 ? items.join("; ") : undefined;
    }
    if (typeof data !== "object") {
        return String(data);
    }
    const errorData = data;
    for (const field of ERROR_MESSAGE_FIELDS) {
        const message = formatResponseErrorData(errorData[field]);
        if (message) {
            return message;
        }
    }
    return stringifyErrorObject(errorData);
};
export const formatServiceHttpErrorMessage = (payload) => {
    const { method, endpoint, status, statusText, data, fallbackMessage } = payload;
    const request = `${method.toUpperCase()} ${endpoint}`;
    const statusMessage = status ? ` responded with ${status}${statusText ? ` ${statusText}` : ""}` : " failed";
    const details = formatResponseErrorData(data) || fallbackMessage;
    return `Allure service request failed: ${request}${statusMessage}${details ? `: ${details}` : ""}`;
};
export const createServiceHttpClient = (serviceUrl, params) => {
    const client = axios.create({
        baseURL: serviceUrl,
        validateStatus: (status) => status < 400,
    });
    const sendRequest = (method) => async (endpoint, payload) => {
        const headers = {
            ...(payload?.headers ?? {}),
        };
        if (params?.accessToken) {
            headers.Authorization = `Bearer ${params.accessToken}`;
        }
        if (params?.apiToken) {
            headers.Authorization = `api-token ${params.apiToken}`;
        }
        try {
            let res;
            if (method === "get" || method === "delete") {
                res = await client[method](endpoint, {
                    ...payload,
                    headers,
                });
            }
            else {
                res = await client[method](endpoint, payload?.body, {
                    ...payload,
                    headers,
                });
            }
            return res.data;
        }
        catch (err) {
            const axiosError = isAxiosError(err);
            if (!axiosError) {
                throw err;
            }
            const { data, status, statusText } = err.response ?? {};
            const responseStatus = status ?? 500;
            const errorMessage = formatServiceHttpErrorMessage({
                method,
                endpoint,
                status,
                statusText,
                data,
                fallbackMessage: err.message,
            });
            if (responseStatus < 500) {
                throw new KnownError(errorMessage, responseStatus);
            }
            throw new UnknownError(errorMessage, err.stack);
        }
    };
    return {
        get: sendRequest("get"),
        post: sendRequest("post"),
        put: sendRequest("put"),
        delete: sendRequest("delete"),
    };
};
const uploadWithRetry = async (filename, uploadAbortController, failedUploads, maxAttempts, maxSimultaneousFailures, uploadFn) => {
    for (let attempt = 1; attempt <= maxAttempts; attempt++) {
        if (uploadAbortController.signal.aborted) {
            return false;
        }
        try {
            await uploadFn();
            failedUploads.delete(filename);
            return true;
        }
        catch (error) {
            if (uploadAbortController.signal.aborted) {
                return false;
            }
            failedUploads.add(filename);
            if (failedUploads.size > maxSimultaneousFailures || attempt >= maxAttempts) {
                throw error;
            }
        }
    }
    return false;
};
export const uploadReport = async (payload) => {
    const { reportUuid, pluginId, files, onProgress, addReportAsset, addReportFile, uploadConcurrency, uploadMaxAttempts, uploadMaxSimultaneousFailures, } = payload;
    const fileEntries = Object.entries(files);
    if (fileEntries.length === 0) {
        return {
            hrefs: {},
        };
    }
    const uploadAbortController = new AbortController();
    const failedUploads = new Set();
    const hrefs = {};
    let indexHref;
    let nextFileIndex = 0;
    const uploadNext = async () => {
        while (!uploadAbortController.signal.aborted) {
            const fileIndex = nextFileIndex++;
            if (fileIndex >= fileEntries.length) {
                return;
            }
            const [filename, filepath] = fileEntries[fileIndex];
            let fileUrl;
            const uploaded = await uploadWithRetry(filename, uploadAbortController, failedUploads, uploadMaxAttempts, uploadMaxSimultaneousFailures, async () => {
                if (isReportDataFile(filename)) {
                    fileUrl = await addReportFile({
                        reportUuid,
                        pluginId,
                        filename,
                        filepath,
                        signal: uploadAbortController.signal,
                    });
                }
                else {
                    await addReportAsset({
                        filename,
                        filepath,
                        signal: uploadAbortController.signal,
                    });
                }
            });
            if (!uploaded || uploadAbortController.signal.aborted) {
                return;
            }
            if (fileUrl) {
                hrefs[filename] = fileUrl;
                if (filename === "index.html") {
                    indexHref = fileUrl;
                }
            }
            onProgress?.();
        }
    };
    const uploadTasks = Array.from({ length: Math.min(uploadConcurrency, fileEntries.length) }, () => uploadNext());
    try {
        await Promise.all(uploadTasks);
    }
    catch (error) {
        uploadAbortController.abort();
        await Promise.allSettled(uploadTasks);
        throw error;
    }
    return {
        indexHref,
        hrefs,
    };
};
